docs/design/llvm_sparse_runtime.md

```diff
@@ -0,0 +1,458 @@
+
+```cpp
+void ListManager::touch_chunk(int chunk_id) {
+  taichi_assert_runtime(runtime, chunk_id < max_num_chunks,
+                        "List manager out of chunks.");
+  if (!chunks[chunk_id]) {
+    locked_task(&lock, [&] {
+      // may have been allocated during lock contention
+      if (!chunks[chunk_id]) {
+        grid_memfence();
+        auto chunk_ptr = runtime->request_allocate_aligned(
+            max_num_elements_per_chunk * element_size, 4096);
+        atomic_exchange_u64((u64 *)&chunks[chunk_id], (u64)chunk_ptr);
+      }
+    });
+  }
+}
+```
+
+By now, you should be quite familiar with this double-checked locking pattern. One thing noteworthy is that the fundamental memory allocation function is provided by [`LLVMRuntime::request_allocate_aligned()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L846-L866).
+
+[`get_element_ptr()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L480-L483) illustrates the way to lookup the address of an element from a given index:
+
+```cpp
+Ptr get_element_ptr(i32 i) {
+  return chunks[i >> log2chunk_num_elements] +                       // chunk base
+          element_size * (i & ((1 << log2chunk_num_elements) - 1));  // slot within the chunk
+}
+```
+
+There is also a reverse operation, [`ptr2index()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L499-L509). It iterates over each chunk to see if the passed-in `ptr` falls into that memory range.
+
+With these primitive helpers, [`allocate()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1591-L1594) is trivial to implement:
+
+```cpp
+Ptr ListManager::allocate() {
+  auto i = reserve_new_element();
+  return get_element_ptr(i);
+}
+```
+
+So is [`append()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1586-L1589):
+
+```cpp
+void ListManager::append(void *data_ptr) {
+  auto ptr = allocate();
+  std::memcpy(ptr, data_ptr, element_size);
+}
+```
+
+[`clear()`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L472-L474) simply resets `num_elements` to zero, without doing anything to the list contents.
+
+## GC
+
+The GC process for a given SNode is divided into three stages:
+
+1. [`gc_parallel_0`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1600-L1626): Moves the remaining, unused indices in `free_list` to its head.
```

Comment:
```suggestion
1. [`gc_parallel_0`](https://github.com/taichi-dev/taichi/blob/172cab8a57fcfc2d766fe2b7cd40af669dadf326/taichi/runtime/llvm/runtime.cpp#L1600-L1626): Moves the remaining, unused indices in `free_list` to its head.

A little bit of context: before GC happens, part of the `free_list` may be already reused. Since `free_list` is a [double-ended queue](https://en.wikipedia.org/wiki/Double-ended_queue), this means the head of the queue is non-zero. To prevent `free_list` from gaining an unlimited length and consume too much memory space as the program runs, we need to move the remaining free cells to the beginning of the queue.
```

(Off-topic: if we use a stack instead of deque for `free_list`, can we save `gc_parallel_0`? )

Reply:
I believe so!